bgpd/bgp_route.c

```diff
@@ -4539,18 +4541,195 @@ static void bgp_soft_reconfig_table(struct peer *peer, afi_t afi, safi_t safi,
 		}
 }
 
+/* Do soft reconfig table on a part of bgp table.
+ * Walk on SOFT_RECONFIG_THREAD_MAX_PREFIX nodes
+ * and schedule a new thread to continue the job.
+ * Without splitting the full job into several part,
+ * vtysh waits for the job to finish before responding to a BGP command
+ */
+static int bgp_soft_reconfig_table_thread(struct thread *thread)
+{
+	struct soft_reconfig_table *srta, *tmp_srta;
+	uint32_t iter, max_iter;
+	int ret;
+	struct bgp_dest *dest;
+	struct bgp_adj_in *ain;
+	struct peer *peer;
+	afi_t afi;
+	safi_t safi;
+	struct bgp_table *table;
+	struct prefix_rd *prd;
+	struct listnode *node, *nnode;
+	bool srta_found;
+
+	srta = THREAD_ARG(thread);
+	peer = srta->peer;
+	afi = srta->afi;
+	safi = srta->safi;
+	table = srta->table;
+	prd = srta->prd;
+	dest = srta->dest;
+
+	if (!table)
+		table = peer->bgp->rib[afi][safi];
+
+	max_iter = SOFT_RECONFIG_THREAD_MAX_PREFIX;
+	if (!srta->dest) {
+		/* first call of the function with a new srta structure.
+		 * Don't do any treatment this timeon nodes
+		 * in order vtysh to respond quickly
+		 */
+		max_iter = 0;
+	}
+
+	for (iter = 0; (dest && iter <= max_iter);
+	     dest = bgp_route_next(dest)) {
+		srta_found = false;
+		for (ALL_LIST_ELEMENTS(dest->soft_reconfig_table, node, nnode,
+				       tmp_srta)) {
+			if (tmp_srta != srta)
+				continue;
+			srta_found = true;
+		}
+		if (!srta_found)
+			continue;
+
+		listnode_delete(dest->soft_reconfig_table, srta);
+		iter++;
+
+		for (ain = dest->adj_in; ain; ain = ain->next) {
+			if (ain->peer != peer)
+				continue;
+
+			struct bgp_path_info *pi;
+			uint32_t num_labels = 0;
+			mpls_label_t *label_pnt = NULL;
+			struct bgp_route_evpn evpn;
+
+			for (pi = bgp_dest_get_bgp_path_info(dest); pi;
+			     pi = pi->next)
+				if (pi->peer == peer)
+					break;
+
+			if (pi && pi->extra)
+				num_labels = pi->extra->num_labels;
+			if (num_labels)
+				label_pnt = &pi->extra->label[0];
+			if (pi)
+				memcpy(&evpn, &pi->attr->evpn_overlay,
+				       sizeof(evpn));
+			else
+				memset(&evpn, 0, sizeof(evpn));
+
+			ret = bgp_update(peer, bgp_dest_get_prefix(dest),
+					 ain->addpath_rx_id, ain->attr, afi,
+					 safi, ZEBRA_ROUTE_BGP,
+					 BGP_ROUTE_NORMAL, prd, label_pnt,
+					 num_labels, 1, &evpn);
+
+			if (ret < 0) {
+				bgp_dest_unlock_node(dest);
+				listnode_delete(peer->bgp->soft_reconfig_table,
```

Comment:
bgp_update() can return -1 in case maximum prefix is reached.
not sure it can happen upon soft reconfigure.

I understand we remove srta for bgp->soft_reconfig_table. ( because we will not iterate again).

But in this situation, I think we should also listnode_delete() the entries for remaining bgp_dest associated.

the order would be:
- bgp_dest_unlock_node()
- for all bgp_dest from this table with same srta, listnode_delete
- listnode_delete(peer->bgp)

Reply:
OK https://github.com/FRRouting/frr/pull/7457/commits/8f520cbc0182a3bb4936d606afc4631a5ab63ccb